ipatests/test_integration/test_epn.py

```diff
@@ -0,0 +1,215 @@
+#
+
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+from __future__ import print_function, absolute_import
+
+import datetime
+import json
+import logging
+import pytest
+
+from ipatests.test_integration.base import IntegrationTest
+from ipatests.pytest_ipa.integration import tasks
+
+logger = logging.getLogger(__name__)
+
+
+def datetime_to_generalized_time(dt):
+    """Convert datetime to LDAP_GENERALIZED_TIME_FORMAT
+       Note: Move into ipalib.
+    """
+    dt = dt.timetuple()
+    generalized_time_str = str(dt.tm_year) + "".join(
+        "0" * (2 - len(str(item))) + str(item)
+        for item in (dt.tm_mon, dt.tm_mday, dt.tm_hour, dt.tm_min, dt.tm_sec,)
+    )
+    return generalized_time_str + "Z"
+
+
+def configure_postfix(host):
+    """Configure postfix to be the destination of the IPA domain.
+    """
+    host.run_command(["systemctl", "start", "postfix"])
+    result = host.run_command(["postconf", "mydestination"])
+    mydestination = result.stdout_text.strip() + ", " + host.domain.name
+    cmd = ["postconf", "-e", mydestination]
+    print(cmd)
+    host.run_command(cmd)
+
+
+class TestEPN(IntegrationTest):
+    """Test Suite for EPN: https://pagure.io/freeipa/issue/3687
+    """
+
+    num_clients = 1
+
+    def _check_epn_output(
+        self,
+        host,
+        dry_run=False,
+        from_nbdays=None,
+        to_nbdays=None,
+        raiseonerr=True,
+    ):
+        result = tasks.ipa_epn(host, raiseonerr=raiseonerr, dry_run=dry_run,
+                               from_nbdays=from_nbdays,
+                               to_nbdays=to_nbdays)
+        json.dumps(json.loads(result.stdout_text), ensure_ascii=False)
+        return (result.stdout_text, result.stderr_text)
+
+    @classmethod
+    def install(cls, mh):
+        tasks.install_master(cls.master, setup_dns=True)
+        tasks.install_client(cls.master, cls.clients[0])
+        # Work around https://github.com/freeipa/freeipa-pr-ci/issues/363
+        # and linting via this broad exception:
+        try:
+            tasks.install_packages(cls.master, ["postfix"])
+            tasks.install_packages(cls.clients[0], ["postfix"])
+            configure_postfix(cls.master)
+            configure_postfix(cls.clients[0])
+        except Exception:
+            # postfix isn't yet used so this shouldn't cause existing
+            # tests to fail.
+            pass
```

Comment:
You might want to do away with this try/except. This was to workaround the PR-CI issue where packages weren't being installed properly.

Reply:
Actually PR-CI is not fixed yet.
Failed run without the try/except:
http://freeipa-org-pr-ci.s3-website.eu-central-1.amazonaws.com/jobs/983787d0-a703-11ea-bfc3-fa163e401a00/report.html 
There is a proposed fix from @wladich in:
https://github.com/freeipa/freeipa-pr-ci/issues/363

Reply:
@fcami My PR https://github.com/freeipa/freeipa-pr-ci/issues/363 is not about fixing dnf issue, it is about fixing other issue without breaking the dnf.
The dnf issue was expected to be resolved by https://github.com/freeipa/freeipa-pr-ci/pull/367

Reply:
@wladich argh, I mixed things up. Do you know if the current runners are up to date?
Because that's [what I get](http://freeipa-org-pr-ci.s3-website.eu-central-1.amazonaws.com/jobs/983787d0-a703-11ea-bfc3-fa163e401a00/report.html) (on a client):

```
[ipatests.pytest_ipa.integration.host.Host.master.cmd48] Error: Error downloading packages:
[ipatests.pytest_ipa.integration.host.Host.master.cmd48]   Curl error (6): Couldn't resolve host name for https://mirrors.fedoraproject.org/metalink?repo=updates-released-f32&arch=x86_64 [Could not resolve host: mirrors.fedoraproject.org]
[ipatests.pytest_ipa.integration.host.Host.master.cmd48] Exit code: 1
ipa: ERROR: stderr: Error: Error downloading packages:
  Curl error (6): Couldn't resolve host name for https://mirrors.fedoraproject.org/metalink?repo=updates-released-f32&arch=x86_64 [Could not resolve host: mirrors.fedoraproject.org]
```

Reply:
What's strange is that ipa-healthcheck similarly installs itself during its testing and that is passing. I wonder if the timing of the package install is different between the tests (e.g. before/after ipa-server-install).

Reply:
Good catch... Let me try changing that